File: 算法/代码随想录算法PDF/00调试源代码/707.设计链表.cpp
/*
 * @lc app=leetcode.cn id=707 lang=cpp
 *
 * [707] 设计链表
 */

// @lc code=start
class MyLinkedList {
public:
    // 链表节点
    struct ListNode {
        int val;
        ListNode* next;
        ListNode(int x) : val(x), next(nullptr) {}
    };

    // 初始化链表
    MyLinkedList() {
        _dummyHead = new ListNode(0);
        _size = 0;
    }
    
    // 获取对应下标节点值
    int get(int index) {  // 注意判断，从0开始size是开区间，index==size也是超界
        if (index < 0 || index >= _size) return -1;
        ListNode* cur = _dummyHead->next;
        while (index--) {
            cur = cur->next;
        }
        return cur->val;
    }
    // 头部加
    void addAtHead(int val) {
        ListNode* node = new ListNode(val);
        node->next = _dummyHead->next;
        _dummyHead->next = node;
        _size++;
    }
    // 尾部加
    void addAtTail(int val) {
        ListNode* node = new ListNode(val);
        ListNode* cur = _dummyHead;
        while (cur->next != nullptr) {
            cur = cur->next;
        }
        cur->next = node;
        _size++;
    }
    // 中间加
    void addAtIndex(int index, int val) {
        if (index > _size) return;
        if (index < 0) index = 0;
        ListNode* node = new ListNode(val);
        ListNode* cur  = _dummyHead;
        while (index--) {
            cur = cur->next;
        }
        node->next = cur->next;
        cur->next = node;
        _size++;
    }
    // 中间删
    void deleteAtIndex(int index) {
        if (index < 0 || index >= _size) return;
        ListNode* cur = _dummyHead;
        while (index--) {
            cur = cur->next;
        }
        ListNode* del = cur->next;
        cur->next = cur->next->next;
        delete del;
        _size--;
    }
private: 
    int _size;
    ListNode* _dummyHead;
};

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList* obj = new MyLinkedList();
 * int param_1 = obj->get(index);
 * obj->addAtHead(val);
 * obj->addAtTail(val);
 * obj->addAtIndex(index,val);
 * obj->deleteAtIndex(index);
 */
// @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/707.设计链表.py
#
# @lc app=leetcode.cn id=707 lang=python3
#
# [707] 设计链表
#

# @lc code=start
class ListNode:
    def __init__(self, val = 0) -> None:
        self.val = val
        self.next = None
class MyLinkedList:

    def __init__(self):
        self.dummyHead = ListNode()
        self.size = 0

    def get(self, index: int) -> int:
        if index < 0 or index >= self.size:
            return -1
        cur = self.dummyHead.next
        while index:
            cur = cur.next
            index -= 1
        return cur.val

    def addAtHead(self, val: int) -> None:
        node = ListNode(val)
        node.next = self.dummyHead.next
        self.dummyHead.next = node
        self.size += 1

    def addAtTail(self, val: int) -> None:
        node = ListNode(val)
        cur = self.dummyHead
        while cur.next:
            cur = cur.next
        cur.next = node
        self.size += 1

    def addAtIndex(self, index: int, val: int) -> None:
        if index > self.size:
            return
        if index < 0:
            index = 0
        node = ListNode(val)
        cur = self.dummyHead
        while index:
            cur = cur.next
            index -= 1
        node.next = cur.next
        cur.next = node
        self.size += 1

    def deleteAtIndex(self, index: int) -> None:
        if index < 0 or index >= self.size:
            return
        cur = self.dummyHead
        while index:
            cur = cur.next
            index -= 1
        cur.next = cur.next.next
        self.size -= 1


# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)
# @lc code=end

